rust/candid/src/de.rs

```diff
@@ -843,3 +843,184 @@ impl<'de, 'a> de::VariantAccess<'de> for Compound<'a, 'de> {
+macro_rules! decode_impl {
+    ( $($id: ident : $typename: ident),* ) => {
+        impl<'a, $( $typename ),*> ArgumentDecoder<'a> for ($($typename,)*)
+        where
+            $( $typename: Deserialize<'a> ),*
+        {
+            fn decode(de: &mut IDLDeserialize<'a>) -> Result<Self> {
+                $(
+                let $id: $typename = de.get_value()?;
+                )*
+
+                Ok(($( $id, )*))
+            }
+        }
+    }
+}
+
+decode_impl!(a: A);
+decode_impl!(a: A, b: B);
+decode_impl!(a: A, b: B, c: C);
+decode_impl!(a: A, b: B, c: C, d: D);
+decode_impl!(a: A, b: B, c: C, d: D, e: E);
+decode_impl!(a: A, b: B, c: C, d: D, e: E, f: F);
+decode_impl!(a: A, b: B, c: C, d: D, e: E, f: F, g: G);
+decode_impl!(a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H);
+decode_impl!(a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I);
+decode_impl!(a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J);
+decode_impl!(
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I,
+    j: J,
+    k: K
+);
+decode_impl!(
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I,
+    j: J,
+    k: K,
+    l: L
+);
+decode_impl!(
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I,
+    j: J,
+    k: K,
+    l: L,
+    m: M
+);
+decode_impl!(
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I,
+    j: J,
+    k: K,
+    l: L,
+    m: M,
+    n: N
+);
+decode_impl!(
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I,
+    j: J,
+    k: K,
+    l: L,
+    m: M,
+    n: N,
+    o: O
+);
+decode_impl!(
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I,
+    j: J,
+    k: K,
+    l: L,
+    m: M,
+    n: N,
+    o: O,
+    p: P
+);
+
+/// Decode a series of arguments, represented as a tuple. There is a maximum of 16 arguments
+/// supported.
+///
+/// Example:
+///
+/// ```
+/// # use candid::Encode;
+/// # use candid::de::decode_args;
+/// let golden1 = 123u64;
+/// let golden2 = "456";
+/// let bytes = Encode!(&golden1, &golden2).unwrap();
+/// let (value1, value2): (u64, String) = decode_args(&bytes).unwrap();
+///
+/// assert_eq!(golden1, value1);
+/// assert_eq!(golden2, value2);
+/// ```
+pub fn decode_args<'a, Tuple>(bytes: &'a [u8]) -> Result<Tuple>
+where
+    Tuple: ArgumentDecoder<'a>,
+{
+    let mut de = IDLDeserialize::new(bytes)?;
+    let res = ArgumentDecoder::decode(&mut de)?;
+    Ok(res)
+}
+
+/// Decode a single argument.
+///
+/// Example:
+///
+/// ```
+/// # use candid::Encode;
+/// # use candid::de::decode_one;
+/// let golden1 = 123u64;
+/// let bytes = Encode!(&golden1).unwrap();
+/// let value1: u64 = decode_one(&bytes).unwrap();
+///
+/// assert_eq!(golden1, value1);
+/// ```
+pub fn decode_one<'a, T>(bytes: &'a [u8]) -> Result<T>
```

Comment:
Non-blocking, but I wonder if we can implement `ArgumentDecoder` for any `Deserialize`, so that we can avoid special case for one argument.

Reply:
Once specialization makes it past unstable, we could use it for exactly that. See https://github.com/rust-lang/rust/issues/31844

Unfortunately the type system is currently limited and trying to do a generic `impl<T> ArgumentDecoder for T` will conflict with the existing implementations that include tuples.